### DISCUSSION TITLE
implicitly bind json  model properties

### DISCUSSION THREAD

**Discussion by NielsvanHoof** [livewire/livewire#4815] [1. Help] [answered]
Hey, Let's say I have the following property on my model:

```
    "components": {
            "banner_alt": {
                "main_title": "Get in touch with us",
                "main_content": "<div>We are ready to assist you with all your questions. Give us a call, send us an e-mail or fill out the form below!</div>"
            },
            "contact_form": {
                "main_title": "Contactform",
                "main_content": "Fill in the form below and we will get back to you as soon as possible."
            },
            "service_page": {
                "zip": 
                "email": 
                "address": 
                "country": 
                "corporate":
                "telephone": 
                "main_title": 
                "street_address": 
            }
        },
```

While working on the edit functionality, I realised I would have to put every possible property inside of the `rules` array ( including future properties)

I currently have this in my `rules` array 

```
  protected array $rules = [
    'page.components.text_grid.*' => ['nullable', 'string'],
    'page.components.banner_alt.*' => ['nullable', 'string'],
    'page.components.usp_row.*' => ['nullable', 'string'],
    'page.components.text_spec.*' => ['nullable', 'string'],
  ];
```

now this does correctly bind the values into my inputs, but when I try editing them I get the following error in this case the banner_alt component:

```
Cannot bind property [page.components.banner_alt.pre_image_intro_text] without a validation rule present in the [$rules] array on Livewire component: [layouts.edit.master].
```

I could do the following 

```
  protected array $rules = [
    'page.components.banner_alt.main_title' => ['nullable', 'string'],
    'page.components.banner_alt.main_content' => ['nullable', 'string'],
  ];
```

But I would have to do this for every possible component and input value (which could be quite a bit), so is it perhaps possible to "lazily" bind them?

**Comment by silnarm** [comment]
I think \* should work fine here.
Your error says `page.components.***banner***.pre_image_intro_text` while your rule says `page.components.***banner_alt***.*`, what's going on there?

**Reply by NielsvanHoof** [reply]
the banner and banner_alt was just a typo, should be fixed now. 

The `*` rule still isn't working, I'll show you the entire class. Not sure what's going wrong.
```
class Master extends Component
{
  public string $layout = '';

  public Page $page;

  protected $listeners = ['setLayout'];

  protected array $rules = [
    'page.components.text_grid.*' => ['nullable', 'string'],
    'page.components.banner.*' => ['nullable', 'string'],
    'page.components.banner_alt.*' => ['nullable', 'string'],
    'page.components.usp_row.*' => ['nullable', 'string'],
    'page.components.text_spec.*' => ['nullable', 'string'],
  ];

  public function mount(): void
  {
    $this->layout = $this->page->page_layout;
  }

  public function render(): Factory|View|Application
  {
    return view('livewire.layouts.edit.master');
  }

  public function setLayout(string $layout): void
  {
    $this->layout = $layout;
  }

  public function updatedState(): void
  {
    $this->emitTo('forms.edit.page.form', 'mergeState', $this->page);
  }
}
```

**Comment by joshhanley** [comment, marked as answer]
@NielsvanHoof I'm guessing `Page` is an eloquent model? If so, I would probably use an array instead of a model, that way you don't need to specify rules for each property you want to access. Then when you are ready to save then process the array into the model.

Hope this helps!

**Reply by NielsvanHoof** [reply]
Would this stil give me the benefit of having inputs pre filled with the model properties? e.g. if I set the `wire:model` to an empty array, I would miss out on having the inputs filled with the model data, or is there any other way this could be achieved ?

**Reply by joshhanley** [reply]
@NielsvanHoof yes, you can get the original model data and convert it to an array to ensure it is pre-filled rather than setting it to an empty array.

Hope this helps!

**Reply by NielsvanHoof** [reply]
Hey, managed to fix it using the `toArray` method, thanks!
